### ISSUE TITLE
ResNet

### ISSUE THREAD

**Issue by JohnCraigPublic** [Microsoft/CNTK#639]
Question: Is the 152-layer ResNet available for use in CNTK?   I heard that it was actually a Caffe model?  But, someone has mapped it into CNTK?   Thanks...

**Comment by Alexey-Kamenev** [comment]
Yes, ResNet-152 is available in CNTK, both config and trained model. Please take a look [here](https://github.com/Microsoft/CNTK/tree/master/Examples/Image/Miscellaneous/ImageNet/ResNet). The original models by K. He et al were trained in Caffe. Shortly after that, CNTK team created and trained similar models in CNTK.

**Comment by JohnCraigPublic** [comment]
OK, maybe a dumb question but....  I followed your link and downloaded ResNet-152 -- what downloaded is a *.model file.  Whats that?   I expected a *.ndl or a *.cntk  -- what do I do with the *.model file?  Thanks.

**Comment by Alexey-Kamenev** [comment]
.model is a binary file which contains a trained model so you don't have to train it yourself. Training ResNet-152 on ImageNet takes about a week on 16 GPUs so training from scratch may not be suitable for many users. For the example on how to use pre-trained models (i.e. .model files), see [evaluation](https://github.com/Microsoft/CNTK/tree/master/Examples/Evaluation) samples.
If you are interested in topology and settings then look at `ResNet_152.cntk` and `ResNet_152.ndl` - these files will allow to train the net from scratch.

**Comment by JohnCraigPublic** [comment]
If I could just ask one or two more questions, I will be ready to roll....
So, I've been using the MNIST example in order to learn CNTK (and next, I want to graduate to ResNet-152) -- but sticking with MNIST right now -- I'm trying to do an Eval from C++ code.
One basic question: after training MNIST with 02_convolution network, there is a file created in MNIST/Output/Models with the name "02_convolution".  Is this the binary model file?  It doesn't have the extension ".model" like the resent model you pointed me to, but its the same sort of file, is that correct?
And the other files left in that folder, with names like "02_convolution.1", "02_convolution.2", etc -- are they temp files that I could delete now?  Or are they needed to eval the model?   Thanks for your help.

**Comment by dongyu888** [comment]
I think "02_convolution.1", "02_convolution.2" are model files after epoch 1, 2, etc. if you don’t need to select which model to use based on some dev set, you can delete them and only evaluate based on the final model file, which in this case, is likely to be 02_convolution

**Comment by JohnCraigPublic** [comment]
and "02_convolution" could be called "02_convolution.model" ?    Or is a ".model" file something else?

**Comment by dongyu888** [comment]
It depends on what is specified as the modelPath in the config file.

**Comment by jbalma** [comment]
Hi John,

The files with extensions like ResNet_152.0 or ResNet_152.3 are what gets saved after each epoch as you train the ResNet_152 network. If you don't want to restart training from where you left off, you can delete these, but if you are training a large, long running model like ResNet_152, you'll probably want to save them. The ResNet_152.cntk and ResNet_152.ndl are given as input to CNTK to describe the network topology, training/validation data, and output parameters. 

Is your goal to train a ResNet model or test a pretrained model? 

If you are trying to train, I would start by verifying you have the following paths specified in the Train, CreateEval, and Test descriptions within the config file ResNet_152.cntk:

In ResNet_152.cntk:
-------------------------------------
  RootDir = "."

  ConfigDir = "$RootDir$"
  DataDir = "$RootDir$"
  OutputDir = "$RootDir$/Output"
  ModelDir = "$OutputDir$/Models"
  ndlMacros="$ConfigDir$/Macros.ndl"

  stderr="$OutputDir$/ResNet_152"
...
Train=[
    action="train"
    modelPath="$ModelDir$/ResNet_152"

     NDLNetworkBuilder=[
        networkDescription="$ConfigDir$/ResNet_152.ndl"
    ]
...
]

CreateEval=[
    action="edit"
    CurModel="$ModelDir$/ResNet_152"
    NewModel="$ModelDir$/ResNet_152.Eval"
    editPath="$ConfigDir$/CreateEvalModel.mel"
]

Test=[
    action="test"
    modelPath="$ModelDir$/ResNet_152.Eval"
    minibatchSize=512

    reader=[
        readerType="ImageReader"
        file="$DataDir$/val_map.txt"
...
]
...
----------------------------------------------


That stderr variable is important because it determines what the output files (from each rank) will be labeled as. For example, if you run ResNet_152 across 16 ranks, you'll see 16 files labeled something like "ResNet_152_Train_CreateEval.logrank0" in the directory ./Output.

Likewise, as the training progresses, you'll see new "model" files appear in the ./Output/Models directory labeled as ResNet_152.7, only now the number indexing the files is the epoch. If the run stops before the max number of epochs is reached, there will also be a checkpoint file saved labeled like ResNet_152.7.ckp, which I think is needed to restart the job from where you left off.

Like Alexey said, training ResNet_152 takes a long time because the network size is very large and restricts the size of the mini-batch and data motion to/from the GPU. If you're just starting with ResNet in CNTK, I'd try something smaller like ResNet_18 with the default MB size to test whether or not it's working with whatever data/hardware you're using, since it runs quite a bit faster than the larger networks. 

You also need to make sure you have provided a "train_map.txt" file (described in which specifies the location of the images you want to use for training and their corresponding label. There should be a similar file for testing "val_map.txt", specified in ResNet_152.cntk (under "Test") as well, using data the network hasn't seen in training. Hope this helps. Good luck!

**Comment by JohnCraigPublic** [comment]
Thanks jbalma!

**Comment by priyathamkat** [comment]
@Alexey-Kamenev  The link to ResNet is broken, though I found it [here](https://github.com/Microsoft/CNTK/tree/master/Examples/Image/Classification/ResNet). But it doesn't have a trained model. Can you please tell me where I can download it?

**Comment by hellosaumil** [comment]
Can anybody please with this #2732 ? Thank you. 

@JohnCraigPublic @Alexey-Kamenev @jbalma @dongyu888 @priyathamkat
